wordpress-mobile/WordPress-FluxC-Android
# [Add Detekt] 1.21.0 Integration

Parent: #2476
Closes: #2477

This PR integrates the [latest stable](https://github.com/detekt/detekt/releases) version of Detekt to the repo ([1.21.0](https://github.com/detekt/detekt/releases/tag/v1.21.0)), and includes the following:

- Uses [android-dependency-catalog](https://github.com/Automattic/android-dependency-catalog) and its [1.3.0](https://github.com/Automattic/android-dependency-catalog/releases/tag/1.3.0) version, which includes [Detekt 1.21.0](https://github.com/Automattic/android-dependency-catalog/pull/5) in it (1e1f82865f7bbe735e8c0cab0cd867d6cf2601c4 + cd3c43d5670e7880e5d71718bf43ae4bc60f8ed2).
- Creates the basic configuration, with `buildUponDefaultConfig = true` and the base `detekt.yml` overriding configuration, as it being replicated from [WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/config/detekt/detekt.yml) and [WCAndroid](https://github.com/woocommerce/woocommerce-android/blob/trunk/config/detekt/detekt.yml), minus any `Compose` configuration as it is not need for this repo (92a5960a80f4196c823db3d2e7f14ee475f9219e).
- Suppresses all existing warnings via the baseline configuration (1b720dd710144ccfa6f27423b571a5e8ee170458).
- Adds CI configuration for Detekt (4d90d8e3769862d17ce25a61f11a38794a91f75e).
- Documents Detekt and its usage (fe66e1cf6ef6dda469af9a959577a68d419600bd).

In addition to the above, this PR also:
- Fine-tunes the `DataClassShouldBeImmutable` rule to ignore any `@Table` annotated data classes (c057c8200e5c49b90f68f8b8921252d3c6c073ad).
- Suppresses the `MagicNumber` warning for a couple of `WellSqlConfig` functions (a14ba7f41ac14a516d68cd44fedc5ce7ad91f582).

PS.1: Both the above addition changes remove `626` unnecessary baseline warnings, leaving it with a bit more than 500 warnings overall (`669` to be precise).
PS.2: I recommend reviewing this PR commit-by-commit.

-----

To test: 

- There is nothing much to test here.
- Verifying that all the CI checks are successful should be enough (and more specifically, the [detekt](https://buildkite.com/automattic/wordpress-fluxc-android/builds/1573#01826956-a910-4ba6-bc28-38596f742120) one).
- However, if you really want to be thorough you could also do the following:
    1. `Build Upon Default Config`
        - Run `./gradlew detekt` and make sure it successfully completes (🟢).
        - Comment out the `ignoreAnnotated: ['Table']` configuration from the `DataClassShouldBeImmutable` rule from within the `detekt.yml` file and run `./gradlew detekt` again. You will now notice that Detekt reports new warnings (⚠️).
        - Add back the `ignoreAnnotated: ['Table']` configuration from the `DataClassShouldBeImmutable` rule from within the `baseline.xml` file and run `./gradlew detekt` again. Detekt should be 🟢 again.
    2. `Baseline Config`
        - Run `./gradlew detekt` and make sure it successfully completes (🟢).
        - Remove an entry from within the `baseline.xml` file and run `./gradlew detekt` again. You will now notice that Detekt reports this warning (⚠️). PS: You can do that per module (see `` sections).
        - Add back the removed entry from within the `baseline.xml` file and run `./gradlew detekt` again. Detekt should be 🟢 again.
